llvm/tools/llvm-cas-object-format/llvm-cas-object-format.cpp

```diff
@@ -639,32 +641,54 @@ void StatCollector::printToOuts(ArrayRef<CASID> TopLevels) {
          << "  => Note: PtrSize  = " << sizeof(void *) << "B\n"
          << "  => Note: Cost     = Count*HashSize + PtrSize*Children + Data\n";
-  StringLiteral HeaderFormat = "{0,-22} {1,+10} {2,+7} {3,+10} {4,+7} {5,+10} "
-                               "{6,+7} {7,+10} {8,+7} {9,+10} {10,+7}\n";
-  StringLiteral Format = "{0,-22} {1,+10} {2,+7:P} {3,+10} {4,+7:P} {5,+10} "
-                         "{6,+7:P} {7,+10} {8,+7:P} {9,+10} {10,+7:P}\n";
-  outs() << llvm::formatv(HeaderFormat.begin(), "Kind", "Count", "", "Parents",
-                          "", "Children", "", "Data (B)", "", "Cost (B)", "");
-  outs() << llvm::formatv(HeaderFormat.begin(), "====", "=====", "",
-                          "=======", "", "========", "", "========", "",
-                          "========", "");
-  auto printInfo = [Format, NumHashBytes, &Totals = this->Totals](
-                       StringRef Kind, ObjectKindInfo Info) {
-    if (!Info.Count)
-      return;
-    auto getPercent = [](double N, double D) { return D ? N / D : 0.0; };
-    size_t Size = Info.getTotalSize(NumHashBytes);
-    outs() << llvm::formatv(
-        Format.begin(), Kind, Info.Count, getPercent(Info.Count, Totals.Count),
-        Info.NumParents, getPercent(Info.NumParents, Totals.NumParents),
-        Info.NumChildren, getPercent(Info.NumChildren, Totals.NumChildren),
-        Info.DataSize, getPercent(Info.DataSize, Totals.DataSize), Size,
-        getPercent(Size, Totals.getTotalSize(NumHashBytes)));
-  };
-  for (StringRef Kind : Kinds)
-    printInfo(Kind, Stats.lookup(Kind));
-  printInfo("TOTAL", Totals);
-
+  if(ComputeStatsCSV) {
+    outs() << "Kind,"
+           << "Count,"
+           << "Parents,"
+           << "Children,"
+           << "Data (B),"
+           << "Cost (B)"
+           << "\n";
+    auto printInfo = [NumHashBytes](
+                         StringRef Kind, ObjectKindInfo Info) {
+      if (!Info.Count)
+        return;
+      size_t Size = Info.getTotalSize(NumHashBytes);
+      outs() << Kind << ", " << Info.Count << ", " << Info.NumParents << ", "
+             << Info.NumChildren << ", " << Info.DataSize << ", " << Size
+             << "\n";
+    };
+    for (StringRef Kind : Kinds)
+      printInfo(Kind, Stats.lookup(Kind));
+    printInfo("TOTAL", Totals);
+  } else {
+    StringLiteral HeaderFormat = "{0,-22} {1,+10} {2,+7} {3,+10} {4,+7} {5,+10} "
+                                 "{6,+7} {7,+10} {8,+7} {9,+10} {10,+7}\n";
+    StringLiteral Format = "{0,-22} {1,+10} {2,+7:P} {3,+10} {4,+7:P} {5,+10} "
+                           "{6,+7:P} {7,+10} {8,+7:P} {9,+10} {10,+7:P}\n";
+    outs() << llvm::formatv(HeaderFormat.begin(), "Kind", "Count", "", "Parents",
+                            "", "Children", "", "Data (B)", "", "Cost (B)", "");
+    outs() << llvm::formatv(HeaderFormat.begin(), "====", "=====", "",
+                            "=======", "", "========", "", "========", "",
+                            "========", "");
```

Comment:
You've duplicated everything, which means any time we change the stats we're going to have to remember to change two places.

It seems better to avoid nesting, and instead just customize the format. I think something like this would work:
```
  StringLiteral HeaderFormatPretty = "{0,-22} {1,+10} {2,+7} {3,+10} {4,+7} {5,+10} "
                                "{6,+7} {7,+10} {8,+7} {9,+10} {10,+7}\n";
  StringLiteral FormatPretty = "{0,-22} {1,+10} {2,+7:P} {3,+10} {4,+7:P} {5,+10} "
                          "{6,+7:P} {7,+10} {8,+7:P} {9,+10} {10,+7:P}\n";
  StringLiteral FormatCSV = "{0}, {1}, {3}, {5}, {7}, {9}\n";

  StringRef HeaderFormat = IsCSV ? FormatCSV : HeaderFormatPretty;
  StringRef Format = IsCSV ? FormatCSV : FormatPretty;
  outs() << llvm::formatv(HeaderFormat.begin(), "Kind", "Count", "", "Parents",
                           "", "Children", "", "Data (B)", "", "Cost (B)", "");
   if (!IsCSV)
     outs() << llvm::formatv(HeaderFormat.begin(), "====", "=====", "",
                             "=======", "", "========", "", "========", "",
                             "========", "");
```

Reply:
Thanks duncan, helped me trim down the duplication a lot